File: framework/languages/java/samples/java/GameQuest/Server/QuestMission/src/main/java/systems/zlink/samples/gamequest/server/questmission/infrastructure/zlink/ZLinkQuestProgressNotifier.java
package systems.zlink.samples.gamequest.server.questmission.infrastructure.zlink;

import org.springframework.stereotype.Component;
import systems.zlink.framework.channels.ZLinkClient;
import systems.zlink.samples.gamequest.server.configuration.SampleNames;
import systems.zlink.samples.gamequest.server.questmission.application.QuestEventProcessor;
import systems.zlink.samples.gamequest.shared.contracts.Messages;

@Component
public final class ZLinkQuestProgressNotifier implements QuestEventProcessor.QuestProgressNotifier {
    private final ZLinkClient channels;

    public ZLinkQuestProgressNotifier(ZLinkClient channels) {
        this.channels = channels;
    }

    @Override
    public boolean notify(String sourceApi, Messages.NotifyQuestProgressReq request) {
        String apiName = "api-b".equals(sourceApi) ? "api-b" : "api-a";
        try {
            Messages.NotifyQuestProgressRes delivered = channels
                .requestToChannel(SampleNames.gameApiActionChannel(apiName), request)
                .timeout(SampleNames.RequestTimeout)
                .await(Messages.NotifyQuestProgressRes.class);
            return delivered != null && delivered.delivered();
        } catch (RuntimeException error) {
            System.err.printf(
                "gamequest mission projection kept while stream notify failed. player=%s%n",
                request.playerId());
            return false;
        }
    }
}

File: framework/languages/java/samples/kotlin/GameQuest/Server/QuestMission/src/main/kotlin/systems/zlink/samples/kotlin/gamequest/server/questmission/infrastructure/zlink/ZLinkQuestProgressNotifier.kt
package systems.zlink.samples.kotlin.gamequest.server.questmission.infrastructure.zlink

import org.springframework.stereotype.Component
import systems.zlink.framework.channels.ZLinkClient
import systems.zlink.samples.kotlin.gamequest.server.configuration.SampleNames
import systems.zlink.samples.kotlin.gamequest.server.questmission.application.QuestEventProcessor
import systems.zlink.samples.kotlin.gamequest.shared.contracts.NotifyQuestProgressReq
import systems.zlink.samples.kotlin.gamequest.shared.contracts.NotifyQuestProgressRes

@Component
class ZLinkQuestProgressNotifier(
    private val channels: ZLinkClient,
) : QuestEventProcessor.QuestProgressNotifier {
    override fun notify(sourceApi: String, request: NotifyQuestProgressReq): Boolean {
        val apiName = if (sourceApi == "api-b") "api-b" else "api-a"
        return try {
            val delivered = channels
                .requestToChannel(SampleNames.gameApiActionChannel(apiName), request)
                .timeout(SampleNames.RequestTimeout)
                .await(NotifyQuestProgressRes::class.java)
            delivered != null && delivered.delivered
        } catch (error: RuntimeException) {
            System.err.printf(
                "gamequest mission projection kept while stream notify failed. player=%s%n",
                request.playerId,
            )
            false
        }
    }
}
